agent/base

```diff
@@ -293,8 +293,10 @@ function generate_inventory {
 
 function resolve_benchmark_bin {
     # Encapsulation of method for resolving the actual benchmark
-    # binary.
-    which --skip-alias --skip-functions "${1}"
+    # binary ignoring `bash` aliases and functions.
+    unalias "${1}" > /dev/null 2>&1
+    unset -f "${1}" > /dev/null 2>&1
+    command -v "${1}"
```

Comment:
I'm slightly uncomfortable with the issuance of the `unalias` and `unset` commands, here, because they affect the environment of the calling shell.

If `${1}` is an alias, `command` will return `alias` as the command:
```
$ command -v ls
alias ls='ls --color=auto'
```
while if `${1}` is a shell function or built-in, `command` will return the unadorned value of `${1}`:
```
$ command -v which
which
$ command -v echo
echo
```
If `${1}` is anything else (i.e., an executable file), `command` returns a path (containing a slash):
```
$ command -v /bin/echo
/bin/echo
$ command -v exec-unittests 
./exec-unittests
$ command -v whereis
/usr/bin/whereis
```

So, it seems like, rather than modifying the environment (and, perhaps, surprising the user by _silently_ removing an intended redirect), we could instead detect and report the problem.  (I.e., if the value returned by `command -v` starts with `"alias ${1}"` or doesn't contain a slash, then reject it.)  Thoughts?

Reply:
I think you misinterpreted what I meant by "calling shell".

In your example, the `./this` script is being executed in its own shell; therefore, the `unset -f` does not affect the environment in the shell which issued the `source demo` command.

However, in `agent/base`, the `unset -f` is issued inside a shell function, not a subshell, so it affects the environment of the shell which invoked the function...which, as far as I can tell, is very much the intend effect.  So, no, it doesn't affect the parent shell of the `pbench-fio` invocation, but it affects execution within that invocation.

My concern is that it is overriding the user's apparent intention, and it is doing so silently.  That is, if someone tries to set up an alias (or a shell function) for `fio`, it won't work, and there will be no indication of why (and no suggestion that what they needed to do was to change their `PATH`).  And, worse, if there is some other `fio` executable on their path, we'll try to run that instead of the one that the user wanted.

Reply:
But, that is how it worked before.  Aliases did not work before and don't work now. Functions were ignored, and the command was constructed using the resolved path to the actual binary.

This is now only checking for the actual binary, with the side effect of ignoring the function so that the resolved binary is still used.  We _could_ (see below) just document that we explicitly ignore functions.

I am beginning to think we are constructing too fragile a requirement on the "existence" of the executable, and should instead focus on the version check: if it exists, we'll go on to see if it returns a version we support, we're done.  If it doesn't exist, or the version does not match what we expect, we return an error.

Reply:
> But, that is how it worked before.

OK; on that basis, I'm going to go ahead and resolve this conversation.

> I am beginning to think we [...] should instead focus on the version check

I think that that would be great, but IIRC we went this other way because LINPACK didn't have a way to report its version....

Reply:
> I think that that would be great, but IIRC we went this other way because LINPACK didn't have a way to report its version....

`pbench-linpack` doesn't use this method as of today.

Reply:
> `pbench-linpack` doesn't use this method as of today.

I'm not sure which "this method" that is.

My understanding is that `pbench-linpack` doesn't do any version check today, and, in no small part, that's because there is nothing in the LINPACK benchmark workload which produces a version that it can check.  (So, we're trying to manufacture something, by looking at the path to the executable.)  Am I confused?

Reply:
I am a bit confused by what is being discussed.

Checking for the existence of an executable file should be removed (I'll post an update to do that late Monday or Tuesday).

Checking that when the `fio` or `uperf` commands are invoked with the appropriate option to report a version of the program, where that version is acceptable to us, should be the only thing we need to do.  If the invocation to generate a version fails, we should exit with an error.  If the invocation reports a version that we don't support, we should fail.  If it reports a version we do support, we should succeed.

What we do in `pbench-linpack` should not be discussed here since it is handled in an entirely different way and has not bearing on this discussion since the `resolve_benchmark_bin` function provided by `agent/base` is not used there.